File: src/temp/parser.rs
use crate::error::ErrorHandler;
use crate::lexer::Token;
use crate::lexer::TokenType;
use crate::ast::{Program, Stmt, DeclStmt, ActStmt, TuneStmt, VarStmt, PlayStmt, LoopStmt, PlayTuneStmt, AssgnStmt, Expression, BinExpr};

struct Parser {
    tokens: Vec<Token>,
    program: Program,
    current: usize,
    errors: ErrorHandler,
}

fn init_parser(tokens: Vec<Token>, errors: ErrorHandler) -> Parser {
    Parser {
        tokens,
        program: Program {
            statements: Stmt {
                declaration_statements: Vec::new(),
                action_statements: Vec::new(),
            }
        },
        current: 0,
        errors,
    }
}

impl Parser {
    fn parse(&mut self) {
        let (decl_stmt, act_stmt) = self.statement(TokenType::Eof);
        self.program.statements.declaration_statements = decl_stmt;
        self.program.statements.action_statements = act_stmt;
    }

    fn statement(&mut self, until: TokenType) -> (Vec<DeclStmt>, Vec<ActStmt>) {
        let mut decl_stmt: Vec<DeclStmt> = Vec::new();
        let mut act_stmt: Vec<ActStmt> = Vec::new();

        while !self.check_next(until.clone()) && !self.check_next(TokenType::Eof) {
            if self.check_next(TokenType::Tune) {
                decl_stmt.push(self.tune_statement());
            } else if self.check_next(TokenType::Var) {
                decl_stmt.push(self.variable_statement());
            } else if self.check_next(TokenType::Loop) {
                act_stmt.push(self.loop_statement());
            } else if self.check_next(TokenType::Play) {
                act_stmt.push(self.play_statement());
            } else if self.check_next(TokenType::Identifier) {
                act_stmt.push(self.assignment_statement());
            }
        }

        (decl_stmt, act_stmt)
    }

    fn tune_statement(&mut self) -> DeclStmt {
        let token = self.advance(); // Capture The tune Token

        let identifier = self.advance();
        if identifier.ttype != TokenType::Identifier {
            self.new_error("Tune Names Must Be Identifiers", identifier.line);
        }

        let mut parameters: Vec<Token> = Vec::new();

        if self.check_next(TokenType::Colon) {
            self.advance(); // Advance Past The Colon

            while !self.check_next(TokenType::LeftBrace) && !self.check_next(TokenType::Eof) {
                parameters.push(self.advance());
            }
        }

        if !self.check_next(TokenType::LeftBrace) {
            self.new_error("An Opening Brace Was Expected After The Tune Identifier", identifier.line);
        } else {
            self.advance(); // Advance Past The Opening Left Brace
        }

        let (decl_stmt, act_stmt) = self.statement(TokenType::RightBrace);

        if !self.check_next(TokenType::RightBrace) {
            let line = self.peek().line;
            self.new_error("A Closing Brace Was Expected After The Tune Block", line);
        } else {
            self.advance(); // Advance Past The Closing Right Brace
        }

        DeclStmt::TuneStatement(
            TuneStmt {
                token,
                identifier,
                parameters,
                declaration_statements: decl_stmt,
                action_statements: act_stmt,
            }
        )
    }

    fn variable_statement(&mut self) -> DeclStmt {
        let token = self.advance(); // Capture The var Token

        let identifier = self.advance();
        if identifier.ttype != TokenType::Identifier {
            self.new_error("Variable Names Must Be Identifiers", identifier.line);
        }

        if !self.check_next(TokenType::Equal) {
            self.new_error("An Equals Sign Was Expected After The Variable Name", identifier.line);
        } else {
            self.advance(); // Advance Past The Equals Sign
        }

        let value = self.expression();

        if !self.check_next(TokenType::Semicolon) {
            self.new_error("A Semicolon Was Expected After The Variable Declaration", token.line);
        } else {
            self.advance(); // Advance Past The Semicolon
        }

        DeclStmt::VariableStatement(
            VarStmt {
                token,
                identifier,
                value,
            }
        )
    }

    fn loop_statement(&mut self) -> ActStmt {
        let token = self.advance(); // Capture The loop Token

        if !self.check_next(TokenType::Colon) {
            self.new_error("A Colon Was Expected After The Loop Keyword", token.line);
        } else {
            self.advance(); // Advance Past The Colon
        }

        let iterations = self.expression();

        if !self.check_next(TokenType::LeftBrace) {
            self.new_error("An Opening Brace Was Expected After The Iterations Expression", token.line);
        } else {
            self.advance(); // Advance Past The Opening Left Brace
        }

        let (decl_stmt, act_stmt) = self.statement(TokenType::RightBrace);

        if !self.check_next(TokenType::RightBrace) {
            let line = self.peek().line;
            self.new_error("A Closing Brace Was Expected After The Loop Block", line);
        } else {
            self.advance(); // Advance Past The Closing Right Brace
        }

        ActStmt::LoopStatement(
            LoopStmt {
                token,
                iterations,
                declaration_statements: decl_stmt,
                action_statements: act_stmt,
            }
        )
    }

    fn play_statement(&mut self) -> ActStmt {
        if self.peek_next().ttype == TokenType::Colon {
            self.play_note_stmt()
        } else {
            self.play_tune_stmt()
        }
    }

    fn play_note_stmt(&mut self) -> ActStmt {
        let token = self.advance(); // Capture The play Token

        if !self.check_next(TokenType::Colon) {
            self.new_error("A Colon Was Expected After The Play Keyword", token.line);
        } else {
            self.advance(); // Advance Past The Colon
        }

        let note = self.expression();
        let duration = self.expression();
        let velocity = self.expression();

        if !self.check_next(TokenType::Semicolon) {
            self.new_error("A Semicolon Was Expected After The Note, Duration, And Velocity Expressions", token.line);
        } else {
            self.advance(); // Advance Past The Semicolon
        }

        ActStmt::PlayStatement(
            PlayStmt {
                token,
                note,
                duration,
                velocity,
            }
        )
    }

    fn play_tune_stmt(&mut self) -> ActStmt {
        let token = self.advance(); // Capture The play Token
        let identifier = self.advance();

        let mut arguments: Vec<Expression> = Vec::new();

        if self.check_next(TokenType::Colon) {
            self.advance(); // Advance Past The Colon

            while !self.check_next(TokenType::Semicolon) && !self.check_next(TokenType::Eof) {
                arguments.push(self.expression());
            }
        }

        if !self.check_next(TokenType::Semicolon) {
            self.new_error("A Semicolon Was Expected After The Tune Arguments", token.line);
        } else {
            self.advance(); // Advance Past The Semicolon
        }

        ActStmt::PlayTuneStatement(
            PlayTuneStmt {
                token,
                tune: identifier,
                arguments,
            }
        )
    }

    fn assignment_statement(&mut self) -> ActStmt {
        let identifier = self.advance();

        if !self.check_next(TokenType::Equal) {
            self.new_error("An Equals Sign Was Expected After The Identifier Keyword", identifier.line);
        } else {
            self.advance(); // Advance Past The Equals Sign
        }

        let value = self.expression();

        if !self.check_next(TokenType::Semicolon) {
            self.new_error("A Semicolon Was Expected After The Assignment Value", identifier.line);
        } else {
            self.advance(); // Advance Past The Semicolon
        }

        ActStmt::AssignmentStatement(
            AssgnStmt {
                identifier,
                value,
            }
        )
    }

    fn expression(&mut self) -> Expression {
        let lvalue = self.advance().clone();

        if self.check_next(TokenType::Plus) || self.check_next(TokenType::Minus) || self.check_next(TokenType::Multiply) || self.check_next(TokenType::Divide) {
            let op = self.advance();
            let rvalue = self.advance();

            Expression::BinaryExpression(
                BinExpr {
                    lvalue,
                    operator: op,
                    rvalue,
                }
            )
        } else if lvalue.ttype == TokenType::Identifier {
            return Expression::Identifier(lvalue);
        } else {
            return Expression::Number(lvalue);
        }
    }

    fn peek(&mut self) -> Token {
        self.tokens[self.current].clone()
    }

    fn peek_next(&mut self) -> Token {
        self.tokens[self.current + 1].clone()
    }

    fn check_next(&mut self, ttoken: TokenType) -> bool {
        return if self.peek().ttype == ttoken {
            true
        } else {
            false
        }
    }

    fn advance(&mut self) -> Token {
        self.current = self.current + 1;
        self.tokens[self.current - 1].clone()
    }

    fn new_error(&mut self, msg: &str, line: u32) {
        self.errors.add_error(String::from("Parser Error"), String::from(msg), line);

        // Prevents Ghost Errors And Parsing Breakdowns
        self.errors.display_if_has_errors();
    }
}

pub fn parse(tokens: Vec<Token>, errors: ErrorHandler) -> (Program, ErrorHandler) {
    let mut parser = init_parser(tokens, errors);
    parser.parse();

    (parser.program, parser.errors)
}

File: src/midilang/core/Parser.cpp
#include "Parser.h"


MIDILang::Parser::Parser(std::vector<Token> tokens, MIDILang::ErrorHandler* errors) {
    this->tokens = tokens;
    this->current = 0;
    this->errors = errors;
}

MIDILang::Program* MIDILang::Parser::parse() {
    this->program = programNode();
    return this->program;
}

MIDILang::Program* MIDILang::Parser::programNode() {
    return new Program(statementNode(TK_EOF));
}

MIDILang::Statement* MIDILang::Parser::statementNode(TokenType until) {
    auto* declarationStatements = new std::vector<Declaration*>();
    auto* actionStatements = new std::vector<Action*>();

    while (!check(until) && !check(TK_EOF)) {
        if (matchAdvance(TK_TUNE)) declarationStatements->push_back(tuneNode());
        else if (matchAdvance(TK_VAR)) declarationStatements->push_back(variableDeclarationNode());
        else if (matchAdvance(TK_LOOP)) actionStatements->push_back(loopNode());
        else if (matchAdvance(TK_WHILE)) actionStatements->push_back(whileNode());
        else if (matchAdvance(TK_IF)) actionStatements->push_back(ifNode());
        else if (matchAdvance(TK_PLAY)) actionStatements->push_back(playNode());
        else if (matchAdvance(TK_NOTE)) actionStatements->push_back(noteNode());
        else if (peek().ttype == TK_IDENTIFIER) actionStatements->push_back(assignmentNode());
        else newError("Unexpected Token: " + peek().literal, peek().line);
    }

    return new Statement(declarationStatements, actionStatements);
}

MIDILang::Tune* MIDILang::Parser::tuneNode() {
    Token identifier = advance();
    std::vector<Token> parameters = std::vector<Token>();
    Statement* statements;

    if (identifier.ttype != TK_IDENTIFIER) newError("Tune Names Must Be Identifiers", identifier.line);
    if (check(TK_COLON) && (peekN(1).ttype == TK_LEFT_BRACE)) newError("At Least One Parameter Expected After Colon", identifier.line);

    if (matchAdvance(TK_COLON)) while (!check(TK_LEFT_BRACE) && !check(TK_EOF)) {
        parameters.push_back(advance());
        if (parameters.back().ttype != TK_IDENTIFIER) newError("Tune Parameter Must Be An Identifier", identifier.line);
    }

    if (!matchAdvance(TK_LEFT_BRACE)) newError("An Opening Brace Was Expected Before The Tune Block", identifier.line);
    statements = statementNode(TK_RIGHT_BRACE);
    if (!matchAdvance(TK_RIGHT_BRACE)) newError("A Closing Brace Was Expected After The Tune Block", identifier.line);

    return new Tune(identifier, parameters, statements);
}

MIDILang::VariableDecl* MIDILang::Parser::variableDeclarationNode() {
    Token identifier = advance();
    Expression* value;

    if (identifier.ttype != TK_IDENTIFIER) newError("Tune Names Must Be Identifiers", identifier.line);
    if (!matchAdvance(TK_EQUAL)) newError("An Equals Sign Was Expected After The Variable Name", identifier.line);
    value = expressionNode();
    if (!matchAdvance(TK_SEMICOLON)) newError("A Semicolon Was Expected After The Variable Declaration", identifier.line);

    return new VariableDecl(identifier, value);
}

MIDILang::Loop* MIDILang::Parser::loopNode() {
    Expression* iterations;
    Statement* statements;

    if (!matchAdvance(TK_COLON)) newError("A Colon Was Expected After The Loop Keyword", peek().line);
    iterations = expressionNode();
    if (!matchAdvance(TK_LEFT_BRACE)) newError("An Opening Brace Was Expected Before The Iterations Block", peek().line);
    statements = statementNode(TK_RIGHT_BRACE);
    if (!matchAdvance(TK_RIGHT_BRACE)) newError("A Closing Brace Was Expected After The Loop Block", peek().line);

    return new Loop(iterations, statements);
}

MIDILang::While* MIDILang::Parser::whileNode() {
    Expression* condition;
    Statement* statements;

    if (!matchAdvance(TK_COLON)) newError("A Colon Was Expected After The While Keyword", peek().line);
    condition = expressionNode();
    if (!matchAdvance(TK_LEFT_BRACE)) newError("An Opening Brace Was Expected Before The Iterations Block", peek().line);
    statements = statementNode(TK_RIGHT_BRACE);
    if (!matchAdvance(TK_RIGHT_BRACE)) newError("A Closing Brace Was Expected After The While Block", peek().line);

    return new While(condition, statements);
}

MIDILang::If* MIDILang::Parser::ifNode() {
    Expression* condition;
    Statement* ifStatements;

    Statement* elseStatements = nullptr;
    If* nestedIfStatements = nullptr;

    if (!matchAdvance(TK_COLON)) newError("A Colon Was Expected After The If Keyword", peek().line);
    condition = expressionNode();
    if (!matchAdvance(TK_LEFT_BRACE)) newError("An Opening Brace Was Expected Before The If Block", peek().line);
    ifStatements = statementNode(TK_RIGHT_BRACE);
    if (!matchAdvance(TK_RIGHT_BRACE)) newError("A Closing Brace Was Expected After The If Block", peek().line);
    if (matchAdvance(TK_ELSE)) {
        if (matchAdvance(TK_IF)) {
            nestedIfStatements = ifNode();
        } else {
            if (!matchAdvance(TK_LEFT_BRACE)) newError("An Opening Brace Was Expected Before The Else Block", peek().line);
            elseStatements = statementNode(TK_RIGHT_BRACE);
            if (!matchAdvance(TK_RIGHT_BRACE)) newError("A Closing Brace Was Expected After The Else Block", peek().line);
        }
    }

    return new If(condition, ifStatements, elseStatements, nestedIfStatements);
}

MIDILang::Play* MIDILang::Parser::playNode() {
    Token identifier = advance();
    std::vector<Expression*> arguments = std::vector<Expression*>();

    if (matchAdvance(TK_COLON)) while (!check(TK_SEMICOLON) && !check(TK_EOF)) arguments.push_back(expressionNode());
    if (!matchAdvance(TK_SEMICOLON)) newError("A Semicolon Was Expected After The Tune Arguments", identifier.line);

    return new Play(identifier, arguments);
}

MIDILang::Note* MIDILang::Parser::noteNode() {
    int line = peek().line;

    Expression* note = expressionNode();
    Expression* duration = expressionNode();
    Expression* velocity = expressionNode();

    if (!matchAdvance(TK_SEMICOLON)) newError("A Semicolon Was Expected After The Note, Duration, And Velocity Expressions", peek().line);

    return new Note(note, duration, velocity, line);
}

MIDILang::Assignment* MIDILang::Parser::assignmentNode() {
    Token identifier = advance();
    Expression* value;

    if (!matchAdvance(TK_EQUAL)) newError("An Equals Sign Was Expected Before The Assignment Value", identifier.line);
    value = expressionNode();

    if (!matchAdvance(TK_SEMICOLON)) newError("A Semicolon Was Expected After Assignment Expression", peek().line);
    return new Assignment(identifier, value);
}

MIDILang::Expression* MIDILang::Parser::expressionNode() {
    return equalityNode();
}

MIDILang::Expression* MIDILang::Parser::equalityNode() {
    Expression* expression = comparisonNode();

    while (check(TK_COMPARE_EQUAL) || check(TK_COMPARE_NOT_EQUAL)) {
        Token op = advance();
        Expression* rvalue = comparisonNode();
        expression = new Binary(expression, op.literal, rvalue);
    }

    return expression;
}

MIDILang::Expression* MIDILang::Parser::comparisonNode() {
    Expression* expression = termNode();

    while (check(TK_GREATER) || check(TK_GREATER_EQUAL) || check(TK_LESSER) || check(TK_LESSER_EQUAL)) {
        Token op = advance();
        Expression* rvalue = termNode();
        expression = new Binary(expression, op.literal, rvalue);
    }

    return expression;
}

MIDILang::Expression* MIDILang::Parser::termNode() {
    Expression* expression = factorNode();

    while (check(TK_PLUS) || check(TK_MINUS)) {
        Token op = advance();
        Expression* rvalue = factorNode();
        expression = new Binary(expression, op.literal, rvalue);
    }

    return expression;
}

MIDILang::Expression* MIDILang::Parser::factorNode() {
    Expression* expression = unaryNode();

    while (check(TK_MULTIPLY) || check(TK_DIVIDE) || check(TK_MODULUS)) {
        Token op = advance();
        Expression* rvalue = unaryNode();
        expression = new Binary(expression, op.literal, rvalue);
    }

    return expression;
}

MIDILang::Expression* MIDILang::Parser::unaryNode() {
    if (check(TK_NOT) || check(TK_MINUS)) {
        Token op = advance();
        Expression* rvalue = unaryNode();
        return new Unary(op.literal, rvalue);
    } else if (matchAdvance(TK_PLUS)) {
        return unaryNode();
    } else {
        return primaryNode();
    }
}

MIDILang::Expression* MIDILang::Parser::primaryNode() {
    if (check(TK_IDENTIFIER)) {
        return new VariableExpr(advance());
    } else if (check(TK_NUMBER)) {
        return new Literal(advance());
    } else if (matchAdvance(TK_LEFT_PAR)) {
        Expression* expression = expressionNode();
        if (!matchAdvance(TK_RIGHT_PAR)) newError("A Closing Parenthesis Was Expected After The Parenthesized Expression", peek().line);
        return new Grouping(expression);
    }
}

MIDILang::Token MIDILang::Parser::peek() {
    return this->tokens[this->current];
}

MIDILang::Token MIDILang::Parser::peekN(int n) {
    return this->tokens[this->current + n];
}

bool MIDILang::Parser::check(TokenType ttype) {
    if (peek().ttype == ttype) return true;
    else return false;
}

MIDILang::Token MIDILang::Parser::advance() {
    this->current++;
    return this->tokens[this->current - 1];
}

bool MIDILang::Parser::matchAdvance(TokenType ttype) {
    if (peek().ttype == ttype) {
        advance();
        return true;
    }

    return false;
}

void MIDILang::Parser::newError(std::string message, int line) {
    this->errors->addError("Parser Error", message, line);

    // Prevents Ghost Errors And Parsing Breakdowns
    this->errors->displayIfHasErrors();
}
